doc/developer/design/20240508_subsource_schemas_after_planning.md

```diff
@@ -0,0 +1,136 @@
+    happening.
+2.  Move the API for planning that occurs during purification into a spot more
+    semantically appropriate, e.g. let the `SouceConnection` trait define how
+    its options should be allowed to be used. e.g. we should provide common
+    errors that both `PurificationError` (for new sources) and `AdapterError`
+    (for additional subsources) can wrap to signal that there are issues with
+    source definitions.
+3.  Restructure purification such that we track the names of the sources we've
+    been requested to create, as well as the "upstream" objects to which they
+    refer––i.e. instead of generating `Vec<CreateSubsourceStatement>` we should
+    generate `BTreeMap<SubsourceName, UpstreamObjectName>`.
+4.  Add a fn to the `SourceConnection` trait like:
+
+    ```rust
+    fn get_output_schema_for_name(
+        &self,
+        upstream_object_name: UpstreamObjectName
+    ) -> Result<RelationDesc, Err>
+    ```
+
+    This will let us ask for the given schema of a subsource after the source
+    itself is planned.
+
+    `UpstreamObjectName` in the code is likely to just be `UnresolvedItemName`.
+
+5.  After planning the primary source, plan all subsources by generating
+    `CreateSubourceStatement`s using the name, reference mapping; we'll get the
+    schema for the new objects by looking them up in the `Source`.
+
+### `ALTER SOURCE...ADD SUBSOURCE`
+
+We should also refactor the way that `ALTER SOURCE...ADD SUBSOURCE` works to
+instead describe the data it wishes add to the source, and then merge it into
+the existing source's "external metadata," e.g. the `Vec<PostgresTableDesc>`,
+obviously error checking for duplicate references, etc.
+
+To support this, we can extend the `SourceConnection` trait further to
+describe the kind of transformations we will allow to each type of source.
+For PG and MySQL this is simple to envision as something like:
+
+```rust
+fn merge_external_data<T, O>(
+    &mut self, metadata: Vec<T>, options: Vec<O>
+) -> Result<(), Err>
+```
+
+Then, being able to add subsources to a source will just require implementing a
+merge algorithm for the new metadata and options.
+
+This is a little more fraught with updating Kafka metadata, so I'm not totally
+sure how we should proceed there; this issue's scope is limited to the existing
+source/subsource structure. I don't think this enters any traps that preclude
+generalizing this concept for Kafka.
+
+Once we've merged the new and existing data into the `Source`, we can use the
+same code path we use when creating the initial set of subsources to add the
+subsequent sets.
+
+#### Mind Palace
+
+There's a world in which creating a new source also uses `merge_external_data`;
+the only distinction being that new sources start from the empty state, and
+modified sources start with whatever state they had when they were modified.
+
+I am not firmly committed to this idea, but the gist of it seems sound. If we
```

Comment:
I also agree the gist seems sound but it could get tricky in the implementation! I would vote for a first pass time-boxed attempt at this to see if you can get it to work